module/VuFind/src/VuFind/Marc/MarcCollection.php

```diff
@@ -0,0 +1,158 @@
+<?php
+/**
+ * MARC collection class.
+ *
+ * PHP version 7
+ *
+ * Copyright (C) The National Library of Finland 2021.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2,
+ * as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * @category VuFind
+ * @package  MARC
+ * @author   Ere Maijala <ere.maijala@helsinki.fi>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org/wiki/development Wiki
+ */
+namespace VuFind\Marc;
+
+/**
+ * MARC collection class.
+ *
+ * @category VuFind
+ * @package  MARC
+ * @author   Ere Maijala <ere.maijala@helsinki.fi>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org/wiki/development Wiki
+ */
+class MarcCollection implements \Countable, \Iterator
+{
+    /**
+     * Supported serialization formats
+     *
+     * @var array
+     */
+    protected $serializations = [
+        'ISO2709' => Serialization\Iso2709::class,
+        'MARCXML' => Serialization\MarcXml::class,
+    ];
+
+    /**
+     * Records in the collection
+     *
+     * @var array
+     */
+    protected $records = [];
+
+    /**
+     * Iteration position
+     *
+     * @var int
+     */
+    protected $position = 0;
+
+    /**
+     * Constructor
+     *
+     * @param string $data MARC record collection in MARCXML or ISO2709 format
```

Comment:
While I think this approach should work for most of VuFind's purposes, I wonder if an exclusively string-based solution is too limiting. For example, what if I want to pass a 4GB MARC file to the "deletes" script? As implemented, we would have to load the whole thing into memory, and it would likely fail. It might be nice to have support for streaming data instead (which I believe could be accomplished for both binary and XML MARC, though XML handling would probably require use of the [XMLReader class](https://www.php.net/manual/en/book.xmlreader.php)).

If I'm reading the code correctly, these problems already exist in File_MARC, so I don't think we urgently need to address them in order to move forward with this work... but I'm wondering if we should at least lay some groundwork now so we could add these as enhancements without having to break existing interfaces.

Reply:
I changed the MarcCollection class definition to account for the possibility of streaming load, most importantly removing the `\Countable` support. File_MARC can do streaming for ISO2709 records but not MARCXML.